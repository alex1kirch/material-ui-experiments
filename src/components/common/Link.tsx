import * as React from "react";
import EventListener from "react-event-listener";

export const STATUS = {
    HOVERED: "hovered",
    NORMAL: "normal",
};

interface IState {
    class: string;
}
interface IProps {
    page?: string;
}
export default class Link extends React.Component<IProps, IState> {
    public state = {
        class: STATUS.NORMAL,
    };

    constructor(props: IProps) {
        super(props);
        // tslint:disable-next-line:no-console
        console.log("constructor");
    }

    public shouldComponentUpdate() {
        // tslint:disable-next-line:no-console
        console.log("shouldComponentUpdate");

        return true;
    }

    public componentWillUpdate() {
        // tslint:disable-next-line:no-console
        console.log("componentWillUpdate");
    }

    public componentDidUpdate() {
        // tslint:disable-next-line:no-console
        console.log("componentDidUpdate");
    }

    public componentDidMount() {
        // tslint:disable-next-line:no-console
        console.log("componentDidMount");
    }

    public componentWillUnmount() {
        // tslint:disable-next-line:no-console
        console.log("componentWillUnmount");
    }

    public componentDidCatch() {
        // tslint:disable-next-line:no-console
        console.log("componentDidCatch");
    }

    public onResize() {
        // tslint:disable-next-line:no-console
        console.log("onResize");
    }

    public render() {
        return (
            <a
                className={this.state.class}
                href={this.props.page || "#"}
                onMouseEnter={this.onMouseEnter}
                onMouseLeave={this.onMouseLeave}
            >
                {this.props.children}
                <EventListener target="window" onResize={this.onResize} />
            </a>
        );
    }

    private onMouseEnter = () => {
        this.setState({ class: STATUS.HOVERED });
    };

    private onMouseLeave = () => {
        this.setState({ class: STATUS.NORMAL });
    };
}
