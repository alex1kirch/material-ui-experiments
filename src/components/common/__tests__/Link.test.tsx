import { shallow } from "enzyme";
import enzymeToJson from "enzyme-to-json";
import "jest-enzyme";
import * as React from "react";
import Link, { STATUS } from "../Link";

describe("<Link />", () => {
    xit("renders without crashing", () => {
        const wrapper = shallow(<Link>Facebook</Link>);
        expect(enzymeToJson(wrapper)).toMatchSnapshot();
    });

    xit("should changes the class when hovered", () => {
        const wrapper = shallow(<Link page="http://www.facebook.com">Facebook</Link>);
        expect(wrapper.hasClass(STATUS.NORMAL)).toBeTruthy();

        wrapper.simulate("mouseEnter");
        expect(wrapper.hasClass(STATUS.HOVERED)).toBeTruthy();

        wrapper.simulate("mouseLeave");
        expect(wrapper.hasClass(STATUS.NORMAL)).toBeTruthy();
    });

    xit("should changes the class when hovered (toMatchSnapshot)", () => {
        const wrapper = shallow(<Link page="http://www.facebook.com">Facebook</Link>);
        expect(enzymeToJson(wrapper)).toMatchSnapshot();

        wrapper.simulate("mouseEnter");
        expect(enzymeToJson(wrapper)).toMatchSnapshot();

        wrapper.simulate("mouseLeave");
        expect(enzymeToJson(wrapper)).toMatchSnapshot();
    });

    it("shallow lifecycle", () => {
        let wrapper = shallow(<Link page="http://www.facebook.com">Facebook</Link>);
        wrapper = wrapper.setProps({ page: "111" });
        wrapper.unmount();
        expect(enzymeToJson(wrapper)).toMatchSnapshot();
    });
});
