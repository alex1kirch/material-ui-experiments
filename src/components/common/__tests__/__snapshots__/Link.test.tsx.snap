// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> shallow lifecycle 1`] = `
<a
  className="normal"
  href="111"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;
